crates/biome_js_analyze/src/semantic_analyzers/nursery/use_jsx_key_in_iterable.rs

```diff
@@ -0,0 +1,295 @@
+    /// Check out React documentation for [explanation on the why does React need keys.](https://react.dev/learn/rendering-lists#why-does-react-need-keys)
+    ///
+    /// Source: [Eslint `react/jsx-key`](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-key.md)
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// [<Hello />, <Hello />, <Hello />];
+    /// data.map((x) => <Hello>{x}</Hello>);
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// [<Hello key="first" />, <Hello key="second" />, <Hello key="third" />];
+    /// data.map((x) => <Hello key={x.id}>{x}</Hello>);
+    /// ```
+    ///
+    pub UseJsxKeyInIterable {
+        version: "next",
+        name: "useJsxKeyInIterable",
+        recommended: false,
+    }
+}
+
+declare_node_union! {
+    pub UseJsxKeyInIterableQuery = JsArrayExpression | JsCallExpression
+}
+
+declare_node_union! {
+    pub ReactComponentExpression = JsxTagExpression | JsCallExpression
+}
+
+impl Rule for UseJsxKeyInIterable {
+    type Query = Semantic<UseJsxKeyInIterableQuery>;
+    type State = TextRange;
+    type Signals = Vec<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let model = ctx.model();
+
+        match node {
+            UseJsxKeyInIterableQuery::JsArrayExpression(node) => handle_collections(node, model),
+            UseJsxKeyInIterableQuery::JsCallExpression(node) => {
+                handle_iterators(node, model).unwrap_or_default()
+            }
+        }
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let diagnostic = RuleDiagnostic::new(
+            rule_category!(),
+            state,
+            markup! {
+                "Missing \"key\" prop for element in iterator"
+            },
+        )
+        .note(markup! {
+            "The order of the items may change, and having a key can help React identify which item was moved"
+        }).note(markup! {
+            "Check the "<Hyperlink href="https://react.dev/learn/rendering-lists#why-does-react-need-keys">"React documentation"</Hyperlink>". "
+        });
+        Some(diagnostic)
+    }
+}
+
+/// Handle collections of components
+///
+/// Examples
+///
+/// ```js
+/// [<h1></h1>, <h1></h1>]
+/// ```
+fn handle_collections(node: &JsArrayExpression, model: &SemanticModel) -> Vec<TextRange> {
+    node.elements()
+        .iter()
+        .filter_map(|node| {
+            let node = node.ok()?;
+            // no need to handle spread case, if the spread argument is itself a list it
+            // will be handled during list declaration
+            let node = node.as_any_js_expression()?;
+            let res = node
+                .syntax()
+                .descendants()
```

Comment:
I am not sure if it is intentional: `descendants` traverse the entire subtree of nodes. Same for other occurrences of  `descendants`

Reply:
I want to use `descendants` here cause I want to handle cases like
```javascript
let abc = data.map((x) => {let xyz = <h1></h1>; return xyz}
```
But this can cause some false positive like in this case
```javascript
let abc = data.map((x) => {let xyz = <h1></h1>; return <div>{xyz}</div>}
```
Using `descendants` here will force `<h1>` tag to have a key as well, which is not good.

A possible alternative is to match the callback body for only return statements, but I'm afraid it could lead to some negative.

Reply:
My suggestion would be to look for `return` expressions only, and trigger a diagnostic if the expression is not JSX with a clear message explaining a key *may* not be present. Something like:

```
Cannot determine whether this child has the required `key` prop. Either return a JSX expression, or suppress this instance if you determine it is safe.
```

It would be a bit more strict than the ESLint equivalent, but I've noticed with ESLint it would sometimes miss occurrences that could lead to issues.

Reply:
Got it, thanks for your suggestion

Reply:
I've change the implementation to behave like this:
If the collection or iterator is inside a JSX expression, like
```typescript
<h1>{data.map(x => <h2>{x}</h2>)}</h1>
```
Then it will search for the return expression and emit a diagnostic if the returned value is not JSX.
Else if the collection/iterator is not, it will only check for key attribute if the element is JSX